### PR TITLE
BAAS-28253: Add optional tick tracking at the function level

### PR DESCRIPTION
**Pull request by Calvinnix**

### CODE REVIEW

File: vm.go
Comment (Calvinnix) [inline vm.go]: I'm just nervous about the performance implications of having this tick tracking here, I've added a few pieces that should mitigate. I think we could potentially also add sampling if we see performance is suffering in the benchmarks.

i.e. something like `if tickMetricsEnabled && ticks % 10 == 0`

File: compiler_test.go
Comment (Calvinnix) [inline compiler_test.go]: I created a setter method so that I had a central area that I could add the function tick tracking logic.

File: runtime.go
Comment (Calvinnix) [inline runtime.go]: This method might not be necessary since we likely won't flip metrics on/off during a function run.

File: vm.go
Comment (Calvinnix) [inline vm.go]: undo this

File: vm.go
Comment (Calvinnix) [inline vm.go]: I originally had an optimization to save this value on the prg object, but that was when this was being hit for every tick. I don't think that optimization is necessary anymore.

**Comment by Calvinnix** [comment]
The performance impact of this change doesn't look too bad. ([benchmark-vm](https://parsley.mongodb.com/evergreen/baas_rhel70_benchmark_vm_patch_ea700de44c9470142ba9a702fc0a4ce372887388_65badb4057e85ab3c9aa672d_24_01_31_23_44_01/0/task?bookmarks=0,11323))

File: runtime.go
Comment (Calvinnix) [inline runtime.go]: Perhaps this should be something like "functionTickTrackingDisabled" instead.

File: runtime.go
Comment (arahmanan) [inline runtime.go]: I think this should be disabled by default. I would also rename this to `functionTickTrackingEnabled` so that the default value is `false`

File: runtime.go
Comment (arahmanan) [inline runtime.go]: Can we add a comment for this function? We should also mention that this function is not thread safe.

File: runtime.go
Comment (arahmanan) [inline runtime.go]: do we also need a function to enable it?

**Review by arahmanan — COMMENTED** [review]
I sill need to look through all files, but posting some minor comments for now

File: runtime.go
Comment (Calvinnix) [reply runtime.go]: I originally had a function to enable it but ending up removing because we are [checking the setting each function execution](https://github.com/10gen/baas/pull/12723/files#diff-b4c19cbed8278d92c3d7be77f467fefa2b7e8d4809862184c4506f21cef5dc21R412-R416).

However, that approach may change, I'll go ahead and re-add so we have that option moving forward.

File: vm.go
Comment (arahmanan) [inline vm.go]: [nit] To reduce the diversions from goja, what do you think about renaming this function to something along the lines of `profileTicks` and not include setting `vm.prg`. That'll also allow you to simplify some of this code a bit. i.e.
```golang
func (vm *vm) setPrg(prg *Program) {
	if !vm.r.functionTickTrackingEnabled {
		return
	}

	currentTicks := vm.r.Ticks()
	if vm.prg != nil {
		if vm.r.functionTickMetrics == nil {
			vm.r.functionTickMetrics = make(map[string]uint64)
		}
		function := string(vm.prg.funcName)
		if vm.prg.src != nil {
			function = vm.prg.src.Name() + "_" + function
		}
		vm.r.functionTickMetrics[function] += currentTicks - vm.lastFunctionTicks
	}
	vm.lastFunctionTicks = currentTicks
}
```

File: runtime.go
Comment (LijieZhang1998) [inline runtime.go]: [q] We calculate the ticks before `vm.runTry()` or `vm.run()`where ticks is ticking(`vm.r.ticks++`). Would it return 0 in some case? Correct me. I may be lost somewhere.   
[q] can we add some tests to see if we're able to get the ticks metrics correctly?e.g. creating a vm to run a function and return this metrics. It may be not easy because we need to know the expected ticks(lol).

**Review by LijieZhang1998 — COMMENTED** [review]
This is a good idea to track functions. I have some questions about implementation.

File: vm.go
Comment (arahmanan) [inline vm.go]: [nit] Can we initialize this in the `(r *Runtime) init()` function instead to avoid performing this check for each execution? We can probably initialize it only when `functionTickTrackingEnabled` is true.

File: runtime.go
Comment (Calvinnix) [reply runtime.go]: For some background I had a previous iteration of this that tracked ticks at the [exec call](https://github.com/Calvinnix/goja/blob/3a14fee6743a86e40f546b53dbc4bcaad2763cc9/vm.go#L670-L672) but that came with a ~20-30% performance hit.
```
		} else {
                         // tick tracking occurred here
			vm.prg.code[vm.pc].exec(vm)
		}
```

> [q] We calculate the ticks before vm.runTry() or vm.run()where ticks is ticking(vm.r.ticks++). Would it return 0 in some case? Correct me. I may be lost somewhere.

Yes there may be some scenarios on initial setup where this gets hit and no new ticks have been calculated yet. I think for the run and runTry scenario you are referencing it would probably get handled by the existing vm.prg being nil (i.e. this check in setPrg `if vm.prg != nil {`)

I like your idea to add tests, I originally compared the setPrg solution against the results of the verbose solution I referenced above and did not see any discrepancies, but I'll add a few tests to solidify that check.

Comment (arahmanan) [reply runtime.go]: [q1] yes, I believe in some cases the ticks could be 0, but I believe that's fine. I'll let Calvin confirm that thought :) 

[q2] agreed. Now that we're closer to a solution we should start adding some tests. We don't need to check exact ticks, but we should make sure that the metrics contain the correct functions and that the ticks of an expensive JS function are higher than the ticks of trivial functions. It might also be worth testing that executing a function X times generates X times more ticks than executing the same function once.

File: runtime.go
Comment (arahmanan) [inline runtime.go]: [nit] [Google's style guide](https://google.github.io/styleguide/go/decisions#getters) suggests not including the word "get" for getters. I also think the word Function is redundant here, but I'll leave that up to you.
```suggestion
func (self *Runtime) TickMetrics() map[string]uint64 {
```

File: vm.go
Comment (Calvinnix) [reply vm.go]: I like that!

File: vm.go
Comment (LijieZhang1998) [inline vm.go]: What I mean by the situation of returning 0 is when vm.prg != nil. In an extreme case, in the very beginning, `vm.r.Ticks()` should return 0 because ticks is not ticking yet. `vm.lastFunctionTicks` is also 0. So line 328 will plus 0 at that time.

Comment (Calvinnix) [reply vm.go]: Yeah I think that edge case is ok.

Comment (Calvinnix) [reply vm.go]: I say that because I think I prefer the noop (adding 0) over the logic it would take to avoid that scenario.

Let me know if you disagree though!

File: runtime.go
Comment (Calvinnix) [inline runtime.go:499]: I was going to make this map get initialized per the functionTickTrackingEnabled setting but this actual would cause a problem (as demonstrated by the test). We need to have the vm object in order to enable the tick tracking but we also check if it is enabled when we create the vm object. So I think its better to just always initialize this map.

```
			vm := New()
			if tc.functionTickTrackingEnabled {
				vm.EnableFunctionTickTracking()
			}
```
(^ snippet from the unit test)

File: runtime.go
Comment (arahmanan) [inline runtime.go]: [opt] similar comment here about dropping the word "Function". Also, thoughts on grouping all the new tick functions together? i.e. move the `TickMetrics` down here.

File: vm.go
Comment (arahmanan) [reply vm.go]: I also think this is fine, but @LijieZhang1998 let me know if you disagree/if I'm misunderstanding your comment. We don't care too much about these metrics being 100% accurate. We just want to figure out which are the top 10 or so functions we should be focusing on improving.

File: vm.go
Comment (arahmanan) [inline vm.go]: [mega-nit] function names could have an underscore. Thoughts on going with a period instead?
```suggestion
				function = vm.prg.src.Name() + "." + function
```

File: compiler_test.go
Comment (arahmanan) [inline compiler_test.go]: do we need this?

File: func.go
Comment (arahmanan) [inline func.go]: [q] could we call `vm.profileTicks()` within `vm.pushCtx()`?

File: vm.go
Comment (arahmanan) [inline vm.go]: [nit] let's flip this check and return early instead. That'll reduce indentation and improve the readability of this function. i.e.
```golang 
if !vm.r.functionTickTrackingEnabled {
  return
}
```

File: vm_test.go
Comment (arahmanan) [inline vm_test.go]: these tests will break fairly easily if we compare the exact ticks. We don't really care about the tick values. We just want to make sure that the values are reasonable when compared to one another.

File: vm_test.go
Comment (arahmanan) [inline vm_test.go:713]: [q1]can we add some more tests such as executing executing class functions and multiple nested functions?

[q2] Also, do you mind sharing the metrics that we would generate from baas when using a dependency such as the aws sdk? The easiest way to do this might be to run our dependencies tests in evergreen with the tracking enabled.

File: runtime.go
Comment (Calvinnix) [reply runtime.go]: So the only reason I left "Function" here is "EnableTickTracking" sounds like it is for the main tick counter i.e. vm.r.ticks++... However we could do something like "EnableTickMetricTracking"

I agree on grouping these functions together. I'll do that.

File: vm.go
Comment (Calvinnix) [reply vm.go]: The only reason I went with a _ is because the prg.src.Name() can be a file name so `test.js.functionName` looked a bit weird. I can go back to the period though, I don't have a strong opinion on this.

File: compiler_test.go
Comment (Calvinnix) [reply compiler_test.go]: No I'll remove, this is leftover from changing all of the `vm.prg` sets to use the setPrg function.

File: func.go
Comment (Calvinnix) [reply func.go]: I think that's a great idea! This handles a majority of the scenarios. There was a scenario that I left the vm.profileTicks() call but I'll comment on it after I push.

File: vm_test.go
Comment (Calvinnix) [reply vm_test.go]: I'll just check to make sure that the keys line up and the values are greater than 0

File: runtime.go
Comment (Calvinnix) [inline runtime.go:1540]: Without this we would miss the ticks tracked for the above vm.runTry

```
	ex := vm.runTry(r.vm.ctx)
	if ex == nil {
		result = r.vm.result
	} else {
		err = ex
	}
	if recursive {
		vm.popCtx()
		vm.halt = false
		vm.clearStack()
	} else {
		vm.stack = nil
		vm.profileTicks()
		vm.prg = nil
		vm.setFuncName("")
		r.leave()
	}
```

It could turn out that this isn't necessary because it is captured by the pushCtx/restoreCtx, but I don't think it hurts to have this in place just to make sure we capturing any lingering ticks for the program. Worst case scenario is there are no new ticks and we add 0 for the program.

If we think this clutters the code I can dig more into this and see if it is necessary.

File: vm_test.go
Comment (Calvinnix) [reply vm_test.go:713]: > [q1]can we add some more tests such as executing executing class functions and multiple nested functions?

Yeah absolutely!

> [q2] Also, do you mind sharing the metrics that we would generate from baas when using a dependency such as the aws sdk? The easiest way to do this might be to run our dependencies tests in evergreen with the tracking enabled.

Would that data be in splunk? Or would we need to parse the output of the tests?

File: runtime.go
Comment (Calvinnix) [reply runtime.go:1540]: Yeah without this we would miss a few ticks in a few edge cases, not the end of the world but probably nice to keep around.

image

File: vm_test.go
Comment (Calvinnix) [reply vm_test.go:713]: [Evergreen task](https://spruce.mongodb.com/version/65bd532a1e2d17419332be0c/tasks?sorts=STATUS%3AASC%3BBASE_STATUS%3ADESC) with tick metric tracking hardcoded to true. Will update this thread with the results.

File: runtime.go
Comment (arahmanan) [reply runtime.go:1540]: Keeping it works for me.

File: runtime.go
Comment (arahmanan) [inline runtime.go]: should we add a comment for these exported functions?

File: vm_test.go
Comment (arahmanan) [inline vm_test.go]: [opt] Continuing [this thread](https://github.com/mongodb-forks/goja/pull/115#discussion_r1476267287) here.
ah good point... Yeah I don't have a strong opinion either. Maybe we can go with `::`?

File: vm_test.go
Comment (arahmanan) [reply vm_test.go:713]: Sounds good. I'm curious to see if the current approach makes it easy enough for us to identify which function is consuming the most amount of this. Let me know when this is ready.

File: vm_test.go
Comment (Calvinnix) [reply vm_test.go]: Ooo I really like `::` thanks!!

File: vm_test.go
Comment (Calvinnix) [reply vm_test.go:713]: Hmm for some reason the logs aren't showing the "tick metrics" logs for that evergreen run.

I think my local testing might alleviate some of your concerns though.

```
exports = async function(arg){
  const { Buffer } = require('buffer');
  const loremIpsumText = // long string
  return Buffer.from(loremIpsumText);
};
```

image
[Note that the tickMetricsToLog array is sorted so we are looking at the top 5 tick contributors]

The majority of usage for this Buffer.from example is BufferJS::utf8ToBytes and BufferJS::blitBuffer.

Comment (Calvinnix) [reply vm_test.go:713]: I'll also callout that these tick metrics are flat (not cumulative) in that we are measuring the specific time only for that function.

So in this case we would know that we would want to have a native implementation in Go for utf8ToBytes and probably blitBuffer as well, which is nice because this allows us to not have to potentially waste time porting over the entire Buffer.from implementation.

Comment (Calvinnix) [reply vm_test.go:713]: > Hmm for some reason the logs aren't showing the "tick metrics" logs for that evergreen run.

I have a feeling this is happening because of how tick metrics are aggregated and published asynchronously. It doesn't even look like the cmd/server/main.go executes which is where this asynchronous logic would start.

I went ahead and set this up manually though, here are the results when running against [aws-sdk-v3/s3](https://github.com/10gen/baas/blob/165ae68dc53d625d415347bf1703616f7f291f80/function/execution/vm/dependencies_async/dependencies_async_test.go#L299)

image

Comment (LijieZhang1998) [reply vm_test.go:713]: hmac lol

**Review by LijieZhang1998 — APPROVED** [review]
LGTM 💯

**Comment by arahmanan** [comment]
> The performance impact of this change doesn't look too bad. ([benchmark-vm](https://parsley.mongodb.com/evergreen/baas_rhel70_benchmark_vm_patch_ea700de44c9470142ba9a702fc0a4ce372887388_65badb4057e85ab3c9aa672d_24_01_31_23_44_01/0/task?bookmarks=0,11323))

@Calvinnix just to confirm, these benchmarks are with the ticks profiling enabled, right?

**Review by arahmanan — APPROVED** [review]
LGTM! pending [this question](https://github.com/mongodb-forks/goja/pull/115#issuecomment-1928046167)

**Comment by Calvinnix** [comment]
> > The performance impact of this change doesn't look too bad. ([benchmark-vm](https://parsley.mongodb.com/evergreen/baas_rhel70_benchmark_vm_patch_ea700de44c9470142ba9a702fc0a4ce372887388_65badb4057e85ab3c9aa672d_24_01_31_23_44_01/0/task?bookmarks=0,11323))
> 
> @Calvinnix just to confirm, these benchmarks are with the ticks profiling enabled, right?

That is correct, that benchmark was from when the tick profiling was enabled by default and nothing material has changed with the logic since then.

File: vm_test.go
Comment (Gabri3l) [reply vm_test.go:713]: oh nice work, hello mr Buffer not surprised to see you up there!

File: runtime.go
Comment (Gabri3l) [reply runtime.go:499]: [opt] I would document this in code

File: vm_test.go
Comment (Gabri3l) [inline vm_test.go:598]: [nit] The one test I'm missing here is a function that throws an error. I was playing around with popCtx vs restoreCtx and most of the times profiling from either yields the same results. The only exception was when I was using it inside of the restoreCtx (like you are in this PR) where I noticed different ticks values when my function was throwing an error. Just thought it would be useful to have that case here. Realistically it will pass the test anyway and we won't see much different but profiling from popCtx only was missing out on some ticks.

**Review by Gabri3l — APPROVED** [review]
Nice work here, I just wanted one more test case and we're good to go!

File: vm_test.go
Comment (Calvinnix) [reply vm_test.go:598]: Good point! I added a test to cover that we track ticks for a function that errors. 👍

**Pull request merged**
